src/memory/sqlite_store.rs

```diff
@@ -0,0 +1,225 @@
+    pool: Pool<SqliteConnectionManager>,
+    // only enable this in case of a high contention scenario
+    exponential_backoff: bool,
+    _marker: PhantomData<(Address, Word)>,
+}
+
+const MAX_POOL_SIZE: u32 = 10; // placeholder
+const CREATE_TABLE_SCRIPT: &str = "
+CREATE TABLE IF NOT EXISTS findex_aw_store
+(
+    adr BLOB PRIMARY KEY,
+    word BLOB NOT NULL
+)";
+
+impl<Address: Sync, Word: Sync> SqlMemory<Address, Word> {
+    /// Create a new in-memory database.
+    pub fn in_memory(backoff: Option<bool>) -> Result<Self, SqlMemoryError> {
+        let pool = r2d2::Pool::builder()
+            .max_size(MAX_POOL_SIZE)
+            .build(SqliteConnectionManager::memory())?;
+        pool.get().unwrap().execute(CREATE_TABLE_SCRIPT, [])?;
+        Ok(Self {
+            pool,
+            exponential_backoff: backoff.unwrap_or(false),
+            _marker: PhantomData,
+        })
+    }
+
+    /// Connects to a known DB url using the given URL.
+    pub fn connect(url: &str, backoff: Option<bool>) -> Result<Self, SqlMemoryError> {
+        // SqliteConnectionManager::file is the equivalent of using
+        // `rusqlite::Connection::open` as documented in the function's source
+        // code.
+        let pool = r2d2::Pool::builder()
+            .max_size(MAX_POOL_SIZE)
+            .build(SqliteConnectionManager::file(url))?;
+        pool.get().unwrap().execute(CREATE_TABLE_SCRIPT, [])?;
+
+        Ok(Self {
+            pool,
+            exponential_backoff: backoff.unwrap_or(false),
+            _marker: PhantomData,
+        })
+    }
+}
+
+impl<const ADDRESS_LENGTH: usize, const WORD_LENGTH: usize> MemoryADT
+    for SqlMemory<Address<ADDRESS_LENGTH>, [u8; WORD_LENGTH]>
+{
+    type Address = Address<ADDRESS_LENGTH>;
+    type Error = SqlMemoryError;
+    type Word = [u8; WORD_LENGTH];
+
+    async fn batch_read(
+        &self,
+        addresses: Vec<Self::Address>,
+    ) -> Result<Vec<Option<Self::Word>>, Self::Error> {
+        if addresses.is_empty() {
+            return Ok(Vec::new());
+        }
+        let mut conn = self.pool.get()?;
+
+        // Using a transaction to ensure the read is atomic
+        let tx = conn.transaction()?;
+
+        let query = format!(
+            "SELECT adr, word FROM findex_aw_store WHERE adr IN ({})",
+            (0..addresses.len())
+                .map(|_| "?")
+                .collect::<Vec<&str>>()
+                .join(",")
+        );
+
+        // Convert addresses to SQL parameters
+        let params: Vec<ToSqlOutput> = addresses
+            .iter()
+            .map(|addr| {
+                rusqlite::types::ToSqlOutput::Borrowed(rusqlite::types::ValueRef::Blob(&**addr))
+            })
+            .collect();
+
+        // Execute the query
+        let mut stmt = tx.prepare(&query)?;
+        let mut rows = stmt.query(rusqlite::params_from_iter(params))?;
+
+        // Process all rows
+        let mut result_map = std::collections::HashMap::new();
```

Comment:
Using a map here is just wasting CPU cycles: use a vector instead (SELECT should return values in the correct order).